### DISCUSSION TITLE
.NET 8

### DISCUSSION THREAD

**Discussion by dpavicic** [thomhurst/ModularPipelines#412] [General]
Hey Tom,
First, ModularPipelines are great! After wasting months in Nuke trying to create a semi-complex build system, its quirks, and convoluted architecture, I finally gave up, but then I found ModularPipelines, scraped all Nuke stuff, and got back on track. Great job!

If it's not a problem, I want to suggest updating ModularPipelines to .NET 8. There are not only speed improvements in some areas but also nice stuff like keyed services, etc.

For example, if I want to add PipelineGlobalHook as a keyed service, it cannot be done as the compiler breaks with the message: "System.InvalidOperationException: This service descriptor is keyed. Your service provider may not support keyed services."

I made this extension for keyed global hooks:

```
public static IServiceCollection AddKeyedPipelineGlobalHooks<TGlobalSetup>(this IServiceCollection services, string key)
    where TGlobalSetup: class, IPipelineGlobalHooks
  {
    return services.AddKeyedScoped<IPipelineGlobalHooks, TGlobalSetup>(key);
  }
```

This way, one can get the global hook via DI and get some data from it. For example, in my case, I made a hook that scans the repository and makes the internal graph of solutions, projects, and their properties by parsing .csproj files, categorizing them into normal code, analyzers, tests, etc., so they can be easily retrieved in modules later for cleaning, building, packing and all other mumbo-jumbo.

I tried to build from the source to update it myself but had a hard time running and building as there are some issues with project references and some other things. I will post some info about that, also.

**Comment by thomhurst** [comment]
Hey @dpavicic , really glad you're enjoying ModularPipelines!

I don't want to end support for older versions just yet, as I know a lot of companies and such will sit on LTS versions until the last possible moment.

So are you saying even an explicit Package reference to Microsoft.Extensions.DependencyInjection V8 still results in an issue?

However, regardless of that, I can definitely add NET 8 as a multi target option and make that bring in newer versions of libraries like dependency injection.

As for not being able to build the project locally, I'm really sorry about that. Yeah, if you can give some details on the errors you're getting, we'll try and get that sorted!

**Comment by dpavicic** [comment]
Hey Tom, thanks for the quick reply.
Yep, I have my project set to .NET 8. As I see, the ModularPipelines are set to .NET6 and .NET 7, so adding .NET 8 to the list of frameworks shouldn't be a problem.

But..
I managed to run a build by adding all missing dependencies, removing some modules that were failing, and adding net8 in the Directory.Build.props and set the example project to build on .NET8.
So, there will be some work with .NET 8 support. .NET 8 removed support for BinaryFormatter, which ModularPipelines is using (by the looks of it). 
When .NET8 is set as  a target; the build fails in: `PipelineException.cs` in `protected PipelineException(SerializationInfo info, StreamingContext context) : base(info, context)` method with exception:
"Error SYSLIB0051 : 'Exception.Exception(SerializationInfo, StreamingContext)' is obsolete: 'This API supports obsolete formatter-based serialization. It should not be called or extended by application code.' (https://aka.ms/dotnet-warnings/SYSLIB0051)
0>------- Finished building project: ModularPipelines. Succeeded: False. Errors: 1. Warnings: 0"

Here's the post about it:
https://github.com/dotnet/docs/issues/34893

Hope this helps a bit,
Dinko

**Comment by dpavicic** [comment]
I managed to create fix for .NET8. It builds now correctly. PR is away...

**Comment by thomhurst** [comment]
Thanks for fixing that! Honestly I set up that example solution ages ago and forgot about it, so it's not been cared for very well. I might end up removing it eventually if I flesh out the docs a bit more with good examples. But for now, really appreciate the help.

Just one comment on the analyzer reference from the main csproj file. If we remove that, it means analyzers won't be pulled in when installing the main package, and I don't want that to happen, so have suggested a workaround

**Comment by thomhurst** [comment]
@dpavicic I've pushed out an update multi-targeting .NET 8 and bringing in newer versions of the DI libraries.

When you get a chance, let me know if it fixes your issue!
